Crashlytics/UnitTests/FIRCLSRecordTests.m

```diff
@@ -0,0 +1,227 @@
+  XCTAssertEqual(firstImage.base, 4305256448);
+  XCTAssertEqual(firstImage.size, 1392640);
+
+  // Verify last binary
+  FIRCLSRecordBinaryImage *lastImage = adapter.binaryImages[452];
+  XCTAssertTrue(
+      [lastImage.path isEqualToString:@"/System/Library/Frameworks/Accelerate.framework/Frameworks/"
+                                      @"vImage.framework/Libraries/libCGInterfaces.dylib"]);
+  XCTAssertTrue([lastImage.uuid isEqualToString:@"f4421e9313fa386fbd568035eb1d35be"]);
+  XCTAssertEqual(lastImage.base, 7226896384);
+  XCTAssertEqual(lastImage.size, 86016);
+}
+
+- (void)testRecordMetadataFile {
+  FIRCLSRecordAdapter *adapter =
+      [[FIRCLSRecordAdapter alloc] initWithPath:[FIRCLSRecordTests persistedCrashFolder]];
+
+  // Verify identity
+  XCTAssertTrue([adapter.identity.generator isEqualToString:@"Crashlytics iOS SDK/4.0.0-beta.1"]);
+  XCTAssertTrue([adapter.identity.display_version isEqualToString:@"4.0.0-beta.1"]);
+  XCTAssertTrue([adapter.identity.build_version isEqualToString:@"4.0.0-beta.1"]);
+  XCTAssertEqual(adapter.identity.started_at, 1579796954);
+  XCTAssertTrue([adapter.identity.session_id isEqualToString:@"f6cdb3f99e6f4c20a4d68a08c35e7b36"]);
+  XCTAssertTrue(
+      [adapter.identity.install_id isEqualToString:@"169DB25B-8F1D-4115-8364-3887DA9DE73C"]);
+
+  // Verify host
+  XCTAssertTrue([adapter.host.model isEqualToString:@"iPhone11,8"]);
+  XCTAssertTrue([adapter.host.machine isEqualToString:@"N841AP"]);
+  XCTAssertTrue([adapter.host.os_build_version isEqualToString:@"17C54"]);
+  XCTAssertTrue([adapter.host.os_display_version isEqualToString:@"13.3.0"]);
+  XCTAssertTrue([adapter.host.platform isEqualToString:@"ios"]);
+  XCTAssertTrue([adapter.host.locale isEqualToString:@"en_US"]);
+
+  // Verify application
+  XCTAssertTrue(
+      [adapter.application.bundle_id isEqualToString:@"com.google.crashlytics.app.ios-host"]);
+  XCTAssertTrue([adapter.application.build_version isEqualToString:@"1"]);
+  XCTAssertTrue([adapter.application.display_version isEqualToString:@"1.0"]);
+
+  // Verify executable
+  XCTAssertTrue([adapter.executable.architecture isEqualToString:@"arm64"]);
+  XCTAssertTrue([adapter.executable.uuid isEqualToString:@"0341c4166f253830a94a5698cee7fea7"]);
+  XCTAssertEqual(adapter.executable.base, 4305256448);
+  XCTAssertEqual(adapter.executable.size, 1392640);
+}
+
+- (void)testRecordKeyValueFile {
+  FIRCLSRecordAdapter *adapter =
+      [[FIRCLSRecordAdapter alloc] initWithPath:[FIRCLSRecordTests persistedCrashFolder]];
+  XCTAssertEqual(adapter.keyValues.count, 12);
+
+  // Verify first
+  XCTAssertTrue([adapter.keyValues[0].key isEqualToString:@"com.crashlytics.in-background"]);
+  XCTAssertTrue([adapter.keyValues[0].value isEqualToString:@"0"]);
+
+  // Verify last
+  XCTAssertTrue([adapter.keyValues[11].key isEqualToString:@"com.crashlytics.user-id"]);
+  XCTAssertTrue([adapter.keyValues[11].value
+      isEqualToString:@"test-user-28AE6E09-BC30-4CB3-9FA5-FE06828B8F3C"]);
+}
```

Comment:
What do you think about storing a key-value line with an emoji in it, and then reading it out here to make sure they work?

Reply:
I remove this test case as I am now using FIRCLSFile's decode/encode.

This test case would work:
- (void)testHexDecoding {
  NSString *str = @"Hello world! 👻";
  NSString *encoded =  FIRCLSFileHexEncodeString([str UTF8String]);
  NSString *decoded =  FIRCLSFileHexDecodeString([encoded UTF8String]);
  XCTAssertTrue([str isEqualToString:decoded]);
}